.github/workflows/build_and_test.yaml

```diff
@@ -0,0 +1,78 @@
+name: Build and Test PR
+
+on:
+  pull_request: # Trigger for pull requests.
+    types: [opened, synchronize, reopened]
+  workflow_dispatch: # Allows for manual triggering.
+    inputs:
+      ref:
+        description: "The ref to build and test."
+        required: False
+
+jobs:
+  lint:
+    name: Lint
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v2
+        with:
+          path: src
+          ref: ${{ github.event.inputs.ref || github.ref }}
+          # This makes the merge base available for the C++ linter, so that it
+          # can tell which files have changed.
+
+      - name: Lint
+        uses: ./src/.github/workflows/custom-actions/lint-player
+
+  build_and_test:
+    # Don't waste time doing a full matrix of test runs when there was an
+    # obvious linter error.
+    needs: lint
+    strategy:
+      matrix:
+        os: ["ubuntu-latest", "macos-latest", "windows-latest"]
+        browser: ["ChromeHeadless", "FirefoxHeadless", "EdgeHeadless"]
```

Comment:
I tried without "Headless", and tests pass just fine with plain old "Firefox" on Mac, including integration tests.  You're still using --use-xvfb on Linux, and Mac doesn't seem to need headless mode (haven't tested on Windows, but I assume it's the same).